DOC: np.random documentation cleanup and expansion.

rkern opened the pull request:
I've corrected some errors, filled in some gaps, and extended some of the commentary.

rkern committed:
DOC: np.random documentation cleanup and expansion.

rkern commented on doc/source/reference/random/index.rst:
```diff
@@ -69,45 +65,37 @@ is wrapped with a `~.Generator`.
-  from numpy.random import Generator, PCG64
-  rg = Generator(PCG64())
+  from numpy.random import default_gen
+  rg = default_gen(12345)
```
Oops. This depends on #13840

rkern committed:
DOC: Expand on the recommendations.

charris commented:
Test failure looks unrelated.

wkschwartz commented:
According to https://github.com/numpy/numpy/pull/13829#discussion_r297722159 `mtrand.pyx` is closed for new features. This wasn't obvious to the author of #13829 (the changes to `mtrand.pyx` have since been removed), and maybe this won't be obvious to future contributors either. A comment at the top of `mtrand.pyx` saying, as the referenced comment said, that
> This file is frozen and cannot be changed except for issues relating to compilation, not features or bugs.

would help forestall PRs with unacceptable changes

mattip commented on doc/source/reference/random/parallel.rst:
```diff
@@ -5,59 +5,147 @@ There are three strategies implemented that can be used to produce
+.. _`implements an algorithm`: http://www.pcg-random.org/posts/developing-a-seed_seq-alternative.html
+.. _`suffers if there are too many 0s`: http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/MT2002/emt19937ar.html
+.. _`avalanche properties`: https://en.wikipedia.org/wiki/Avalanche_effect
+.. _`not unique to numpy`: https://www.iro.umontreal.ca/~lecuyer/myftp/papers/parallel-rng-imacs.pdf
```
nice explanation

WarrenWeckesser commented on doc/source/reference/random/index.rst:
```diff
@@ -146,6 +134,8 @@ What's New or Different
 * `~.Generator.random` is now the canonical way to generate floating-point
   random numbers, which replaces `random_sample`, `sample`, and `ranf`. This
   is consistent with Python's `random.random`.
+* All BitGenerators in numpy use `~SeedSequence` to process convert seeds into
```
"process convert" -> "convert"

rkern committed:
DOC: typo

charris commented:
Does this supersede #13675?

rkern commented:
Yes.

bashtage commented on doc/source/reference/random/bit_generators/index.rst:
```diff
@@ -46,26 +48,65 @@ Seeding and Entropy
+  # If the user did not provide a seed, it should return `None`.
+  seed = get_user_seed()
+  ss = SeedSequence(seed)
+  print(f'seed = {ss.entropy}')
```
Is 3.6+ syntax OK?

rkern replied:
No, probably not.

bashtage commented on doc/source/reference/random/legacy.rst:
```diff
@@ -39,7 +35,7 @@ are produced in pairs. It is important to use
    lg = RandomState(mt19937)
 
-   # Identical output
+   # Different output, sorry.
```
Should this block be removed since it doesn't actually show what it was written to show?  If you really wanted to show it, you could use

```
mt19937 = MT19937()
mt19973.state = rs.get_state()
lg = RandomState(mt19937)
```

rkern replied:
Good point. That workaround is worth documenting.

bashtage commented on doc/source/reference/random/parallel.rst:
```diff
@@ -5,59 +5,147 @@ There are three strategies implemented that can be used to produce
+a way that you can provide arbitrary-sized integers or lists of integers.
+`~SeedSequence` will take all of the bits that you provide and mix them
+together to produce however many bits the consuming `~BitGenerator` needs to
+initialize itself.
```
One thing I haven't fully understood is does SeedSequence make any guarantees that a n*32bit seed generated from .spawn() are unique.  For example, if I use SeedSequence().spawn(1000000), will the state values produced from these 100000 instances always be distinct? I understand that it isn't possible to guarantee accross multiple generations, (spawn of spawn), but is there any more to say about within generation?

bashtage commented on doc/source/reference/random/parallel.rst:
```diff
@@ -5,59 +5,147 @@ There are three strategies implemented that can be used to produce
+estimated if one knows the number of streams that you derive. `~SeedSequence`
+hashes its inputs, both the seed and the spawn-tree-path, down to a 128-bit
+pool by default. The probability that there is a collision in
+that pool, pessimistically-estimated ([1]_), will be about :math:`n^2*2^{-128}` where
```
I beleive that for some generators the number of draws influences the probability of an overlap. The value is correct for generators that produce independent streams, but could be a bit misleading in some applications.  I suspect this applies to MT. It does not apply to Philox. Not clear about PCG or SFC.

bashtage commented on doc/source/reference/random/parallel.rst:
```diff
@@ -5,59 +5,147 @@ There are three strategies implemented that can be used to produce
+
+.. [1] The algorithm is carefully designed to eliminate a number of possible
+       ways to collide. For example, if one only does one level of spawning, it
+       is guaranteed that all states will be unique. But it's easier to
```
Ahh, here is the guarantee.   Stream of comments...

bashtage commented on doc/source/reference/random/parallel.rst:
```diff
@@ -5,59 +5,147 @@ There are three strategies implemented that can be used to produce
+       estimate the naive upper bound on a napkin and take comfort knowing
+       that the probability is actually lower.
+
+.. [2] In this calculation, we can ignore the amount of numbers drawn from each
```
Same "ahh" here. Should do a review.

bashtage commented:
LGTM. Minor comments that were mostly already addressed.

mattip committed:
DOC: fix reference links

rkern committed:
Merge pull request #1 from mattip/make-refs

DOC: fix reference links

rkern committed:
DOC: clean up examples.

rkern committed:
DOC: Fix description of PCG64 jump.

Also, make the jump step odd to make it a proper Weyl generator. No, this will
never, ever, ever, ever matter. But it feels nicer.

rkern commented on numpy/random/pcg64.pyx:
```diff
@@ -131,17 +131,17 @@ cdef class PCG64(BitGenerator):
+        golden ratio.
         """
-        step = 0x9e3779b97f4a7c15f39cc0605cedc834
+        step = 0x9e3779b97f4a7c15f39cc0605cedc835
```
I need a LGTM on this one last non-trivial bit, then we can merge this. @bashtage ?

bashtage replied:
Feel free to round it which ever way that you prefer. LGTM.

rkern replied:
Thank you for your indulgence. :-)

rkern committed:
DOC: Use old title of np.random section

rkern commented:
@charris This is ready to go once the last CI finishes.

rkern committed:
DOC: Make sure canonical numpy.random target is the index.

charris commented:
Thanks Robert.
